### DISCUSSION TITLE
FastAPI >0.100.0 / Pydantic V2 interpret list query params in pydantic model as body params

### DISCUSSION THREAD

**Discussion by nfcampos** [fastapi/fastapi#10371] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated
from uuid import UUID
from fastapi import FastAPI, Depends, Query
from pydantic import BaseModel


class QueryParamsWithList(BaseModel):
    """Query Params for listing feedback."""

    run: Annotated[list[UUID] | None, Query()] = None
    hello: str | None = None


app = FastAPI()


@app.get("")
async def read_feedbacks(
    query_params: QueryParamsWithList = Depends(),
) -> None:
    print("query_params", query_params)
```

### Description

Open the browser to /redoc, and notice that `run` gets interpreted as a body param (in a GET request...)

<img width="1264" alt="Screenshot 2023-10-03 at 11 40 52" src="https://github.com/tiangolo/fastapi/assets/56902/2122eba9-c415-4c97-a5ad-3dca97ee157a">

Also see that the exact same code produces the expected query param in fastapi==0.96.1, pydantic==1.10.11

<img width="1264" alt="Screenshot 2023-10-03 at 11 42 41" src="https://github.com/tiangolo/fastapi/assets/56902/8feecc90-aaac-4cdb-9f88-71f2d6536c39">

Note this is not an issue with the JSON Schema generation, just using that as a very obvious way of displaying the issue. If you call `GET /?run=...` it gets ignored too

### Operating System

macOS

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.11.5

**Comment by ebottos94** [comment]
For me this is not the correct way to implement query parameters : https://fastapi.tiangolo.com/tutorial/query-params-str-validations/

Anyway, this seems to work : 

```python
from uuid import UUID
from fastapi import FastAPI, Depends, Query
from pydantic import BaseModel, Field


class QueryParamsWithList(BaseModel):
    """Query Params for listing feedback."""

    run: list[UUID] | None = Field(Query(None))
    hello: str | None = None


app = FastAPI()


@app.get("/")
async def read_feedbacks(
    query_params: QueryParamsWithList = Depends()
) -> None:
    return query_params
```

**Comment by nfcampos** [comment]
This doesnt work, as then the default value I get in the endpoint function is `Query(None)` instead of `None`.

**Reply by ebottos94** [reply]
It works : 

![image](https://github.com/tiangolo/fastapi/assets/100039558/fe56675f-df6f-4143-bc0b-76d8af24a4d0)


![image](https://github.com/tiangolo/fastapi/assets/100039558/02b5b752-52ea-447a-99a4-f664b201b619)

**Reply by nfcampos** [reply]
Oh I dont doubt it works if all you do is pass the params back to Pydantic / fastapi. The issue is that other libraries dont't expect to receive `Query(None)` instead of `None`. eg. psycopg/sqlalchemy, therefore this is not the solution to my issue unfortunately

**Reply by ebottos94** [reply]
Why do you say that None is not the default value from the endpoint ? 

![image](https://github.com/tiangolo/fastapi/assets/100039558/f74dab37-94b5-4f74-a119-4aeceec05945)

![image](https://github.com/tiangolo/fastapi/assets/100039558/91cf1a1c-afe1-45a6-abe0-ebc34ed1f8da)

**Comment by elijahsgh** [comment]
I think I'm having a related issue with a pretty simple use case. Please let me know if this is different.

Standard webfinger implementation:
```
class WebfingerRequest(BaseModel):
    resource: str
    rel: Optional[List[str]] = Field(Query([]))


@app.get("/.well-known/webfinger")
async def webfinger(wfr: Annotated[WebfingerRequest, Depends()]):
    return JSONResponse(wfr.model_dump())
```

This works.

**However, using a "normal" model results in rel=None**
```
class WWebfingerRequest(BaseModel):
    resource: str
    rel: Optional[List[str]] = None


@app.get("/.well-known/webfinger")
async def webfinger(wfr: Annotated[WWebfingerRequest, Depends()]):
    return JSONResponse(wfr.model_dump())
```

In the second example there is no "rel" on swagger and it wants a body.

**Reply by CharlesPerrotMinot** [reply]
Yes, had to use the same workaround.

**Comment by YuriiMotov** [comment, marked as answer]
Defining query parameters with `BaseModel` and `Depends()` wasn't officially supported in FastAPI. It worked somehow with Pydantic V1, but doesn't work well with Pydantic V2.

FastAPI officially supported defining Query parameters using models since `0.115.0`.
The syntax is `query_params: QueryParamsWithList = Query()`.
Docs: https://fastapi.tiangolo.com/tutorial/query-param-models/
